README.md
# Rollbar for Android

<!-- RemoveNext -->
Java library for reporting exceptions, errors, and log messages to [Rollbar](https://rollbar.com).

## Setup ##

### Maven

Install from Maven central:

```
<dependencies>
  <dependency>
    <groupId>com.rollbar</groupId>
     <artifactId>rollbar</artifactId>
     <version>0.2.1</version>
  </dependency>
</dependencies>
```

or

```
compile 'com.rollbar:rollbar-android:0.2.1'
```

### Jar

Download [rollbar-android.jar](https://github.com/rollbar/rollbar-android/releases/latest) and place it in your Android project's `libs` directory.

Add the following line in your custom Application subclass's `onCreate()` to initialize Rollbar:

```java
Rollbar.init(this, "POST_CLIENT_ITEM_ACCESS_TOKEN", "production");
```


Make sure your `AndroidManifest.xml` contains the `android.permission.INTERNET` permission.

## Usage ##

By default the notifier will report all uncaught exceptions to Rollbar.

To report caught exceptions, call `reportException()`:

```java
Rollbar.reportException(new Exception("Test exception")); // default level is "warning"

try {
    amount = Integer.parseInt(data);
} except (NumberFormatException e) {
    // Providing a description for this exception
    Rollbar.reportException(e, "critical", "Unexpected data from server");
}
```

To report your own messages, call `reportMessage()`:

```java
Rollbar.reportMessage("A test message", "debug"); // default level is "info"
```

## Configuration reference ##

The following configuration methods are available:

 * **Rollbar.setPersonData(String id, String username, String email)**

    Sets the properties of the current user (called a "person" in Rollbar parlance) to be sent along with every report.

    Default: all `null`


 * **Rollbar.setReportUncaughtExceptions(boolean report)**

    Sets whether or not to report uncaught exceptions to Rollbar.

    Default: `true`


 * **Rollbar.setIncludeLogcat(boolean includeLogcat)**

    Sets whether or not to include logcat output in reports to Rollbar.

    Note: For devices with API level 15 and lower, you will need to include the `android.permission.READ_LOGS` permission in your app's `AndroidManifest.xml` for logcat collection to work.

    Default: `false`


 * **Rollbar.setDefaultCaughtExceptionLevel(String level)**

    Sets the level caught exceptions are reported as by default.

    Default: `"warning"`


 * **Rollbar.setUncaughtExceptionLevel(String level)**

    Sets the level uncaught exceptions are reported as.

    Default: `"error"`


 * **Rollbar.setSendOnUncaughtException(boolean send)**

    If true, uncaught exceptions will immediately be sent to Rollbar, blocking the process shutdown sequence. If false, the exception will just be saved to disk so that it can be sent next time the process starts.

    Default: `false`


 * **Rollbar.setEndpoint(String endpoint)**

    If you are using Rollbar Enterprise, call `setEndpoint` to route events to your Rollbar instance. (Most users will not need to do this.)

    Default: `"https://api.rollbar.com/api/1/items/"`

* **Rollbar.setItemScheduleDelay(int delay)**

    Sets the delay between retries of sending failed items, in seconds.

    Default: `30`



## Deobfuscation ##

If you use [ProGuard](http://developer.android.com/tools/help/proguard.html) to obfuscate your code in production, reported stack traces will not be very useful in most cases.

Rollbar provides a way for you to upload the `mapping.txt` file generated in the obfuscation process so that stack traces can be deobfuscated as they are reported.

Here is an example cURL command to upload a `mapping.txt` file:

```bash
curl 'https://api.rollbar.com/api/1/proguard' \
  -F access_token=POST_SERVER_ITEM_ACCESS_TOKEN \
  -F version=0.0.10 \
  -F mapping=@path/to/mapping.txt
```

Where `version` matches the `android:versionName` in your app's `AndroidManifest.xml`, corresponding to the version the `mapping.txt` was generated for.

After uploading, any future reported exceptions for the specified version will automatically be deobfuscated using the mapping file.

By default, file names and line numbers are removed by ProGuard. To preserve this information to make debugging easier, add the following to your `proguard-project.txt`:

```
-keepattributes SourceFile,LineNumberTable
```

## Help / Support

If you run into any issues, please email us at `support@rollbar.com`

## Build

1. Define property `sdk.dir` which contains path to Android SDK inside `local.properties`
2. Make sure you have installed the platform specified in `project.properties`
  - E.g. `sdk.dir = /Users/coryvirok/Development/adt-bundle-mac-x86_64-20140702/sdk/`
3. Run `ant` from command line

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
# Change Log

**0.2.1**

- Add `code_version` and `version_name` to the top level JSON of the item. See
  [#33](https://github.com/rollbar/rollbar-android/pull/33).
- Fix some issues with the custom thread class to address potential deadlocks. See
  [#32](https://github.com/rollbar/rollbar-android/pull/32).

**0.2.0**

- Improved resiliency against bad network conditions. See [#31](https://github.com/rollbar/rollbar-android/pull/31)
    - Items that fail to send will now be retried periodically. Developers will no longer need to wait until a user restarts the app.

**0.1.3**
- Add pom.xml with metadata for Maven Central

**0.1.2**
- Added flag to `Rollbar.init()` to enable/disable the registration of the uncaught exception handler. (pr#17)

**0.1.1**
- Added reportMessage(message, level, params) to send extra params with messages. (pr#10)

**0.1.0**
- Fixed build.xml, changed `code_version` parameter. (pr#11)
- `code_version` now properly refers to the [versionName](http://developer.android.com/reference/android/content/pm/PackageInfo.html#versionName) attribute from the manifest file.

**0.0.6**
- Added ability to provide a description for caught exceptions. Usage: [https://github.com/rollbar/rollbar-android#usage](https://github.com/rollbar/rollbar-android#usage)
- Added isInit() method to check to see if Rollbar is initialized.

**0.0.5**
- Exceptions with causing exceptions will now be sent with all the causes chained together in a single item. These multiple exceptions will all appear in the proper order in the Rollbar interface.
- Uncaught exceptions are now saved to disk instead of immediately being reported, to prevent any blocking of the app shutdown process. A new configuration option exists to revert to the old behavior.

**0.0.4**
- Ability to include logcat output for every item reported.
- Ability to set person data.
- Caught exception level now defaults to warning, new configuration options available to change this and uncaught exception level.

**0.0.3**
- New persistence system to support retries of item payloads not sent due to network connectivity issues.

**0.0.2**
- Use JSON objects and arrays instead of HashMaps and ArrayLists for payload construction.
- Set reportUncaughtExceptions to true by default.

**0.0.1**
- Initial release.


src/main/java/com/rollbar/android/Notifier.java
package com.rollbar.android;

import java.io.BufferedReader;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager.NameNotFoundException;
import android.text.TextUtils;
import android.util.Log;

import com.rollbar.android.http.HttpRequestManager;
import com.rollbar.android.http.HttpResponse;
import com.rollbar.android.http.HttpResponseHandler;

public class Notifier {
    private static final String NOTIFIER_VERSION = "0.2.1";
    private static final String DEFAULT_ENDPOINT = "https://api.rollbar.com/api/1/items/";
    private static final String ITEM_DIR_NAME = "rollbar-items";
    private static final String PAYLOAD_ERROR_MSG = "There was an error constructing the JSON payload.";
    private static final String ANDROID = "android";
    private static final String MESSAGE = "message";

    private static final int DEFAULT_ITEM_SCHEDULE_STARTUP_DELAY = 1;
    private static final int DEFAULT_ITEM_SCHEDULE_DELAY = 30;
    private static final int MAX_LOGCAT_SIZE = 100;

    private static int itemCounter = 0;

    volatile private boolean handlerScheduled;

    private final ScheduledExecutorService scheduler;

    private final String accessToken;
    private final String environment;

    private JSONObject personData;
    private String endpoint;
    private boolean reportUncaughtExceptions;
    private boolean includeLogcat;
    private String defaultCaughtExceptionLevel;
    private String uncaughtExceptionLevel;
    private boolean sendOnUncaughtException;
    private int itemScheduleDelay;

    private int versionCode;
    private String versionName;

    private final File queuedItemDirectory;
    private final RollbarThread rollbarThread;


    public Notifier(Context context, String accessToken, String environment, boolean registerExceptionHandler) {
        scheduler = Executors.newSingleThreadScheduledExecutor();

        this.accessToken = accessToken;
        this.environment = environment;

        try {
            String packageName = context.getPackageName();
            PackageInfo info = context.getPackageManager().getPackageInfo(packageName, 0);

            this.versionCode = info.versionCode;
            this.versionName = info.versionName;
        } catch (NameNotFoundException e) {
            Log.e(Rollbar.TAG, "Error getting package info.");
        }


        endpoint = DEFAULT_ENDPOINT;
        reportUncaughtExceptions = true;
        defaultCaughtExceptionLevel = "warning";
        uncaughtExceptionLevel = "error";
        sendOnUncaughtException = false;
        itemScheduleDelay = DEFAULT_ITEM_SCHEDULE_DELAY;

        handlerScheduled = false;

        queuedItemDirectory = new File(context.getCacheDir(), ITEM_DIR_NAME);
        queuedItemDirectory.mkdirs();

        if (registerExceptionHandler) {
            RollbarExceptionHandler.register(this);
        }

        rollbarThread = new RollbarThread(this);
        rollbarThread.start();

        scheduleItemFileHandler(DEFAULT_ITEM_SCHEDULE_STARTUP_DELAY);
    }

    private JSONArray getLogcatInfo() {
        JSONArray log = null;

        int pid = android.os.Process.myPid();

        try {
            Process process = Runtime.getRuntime().exec("logcat -d");

            InputStreamReader isr = new InputStreamReader(process.getInputStream());
            BufferedReader br = new BufferedReader(isr, 8192);

            List<String> lines = new ArrayList<String>();

            String line;
            while ((line = br.readLine()) != null) {
                // Only include the line if the current process's pid is present
                if (line.contains(String.valueOf(pid))) {
                    lines.add(line);
                    if (lines.size() > MAX_LOGCAT_SIZE) {
                        lines.remove(0);
                    }
                }
            }

            log = new JSONArray(lines);
        } catch (IOException e) {
            Log.e(Rollbar.TAG, "Unable to collect logcat info.", e);
        }

        return log;
    }

    private JSONObject buildNotifierData() throws JSONException {
        JSONObject notifier = new JSONObject();
        notifier.put("name", "rollbar-android");
        notifier.put("version", NOTIFIER_VERSION);

        return notifier;
    }

    private JSONObject buildClientData() throws JSONException {
        JSONObject client = new JSONObject();

        client.put("timestamp", System.currentTimeMillis() / 1000);

        // Put version code and name at top level of returned JSON so that it can be referenced
        // in the "advanced search" text field on the /items page at rollbar.com. Properties referenced
        // in this field must be at the top level of the returned JSON.
        client.put("code_version", this.versionCode);
        client.put("name_version", this.versionName);

        // Put version code and name under alternative names, in case user accidentally reverses them.
        client.put("version_code", this.versionCode);
        client.put("version_name", this.versionName);

        JSONObject androidData = new JSONObject();
        androidData.put("phone_model", android.os.Build.MODEL);
        androidData.put("android_version", android.os.Build.VERSION.RELEASE);
        // NOTE use versionName instead of versionCode in order to fix
        //      'Resolving Rollbar Items in Versions' feature https://rollbar.com/blog/post/2013/09/17/resolving-rollbar-items-in-versions
        androidData.put("code_version", this.versionName);
        // NOTE keep original versionCode value with different name
        androidData.put("version_code", this.versionCode);
        androidData.put("version_name", this.versionName);

        if (includeLogcat) {
            androidData.put("logs", getLogcatInfo());
        }

        client.put(ANDROID, androidData);
        client.put("user_ip", "$remote_ip");

        return client;
    }

    private JSONObject buildData(String level, JSONObject body) throws JSONException {
        JSONObject data = new JSONObject();

        data.put("environment", this.environment);
        data.put("level", level);
        data.put("platform", ANDROID);
        data.put("framework", ANDROID);
        data.put("language", "java");

        data.put("body", body);

        if (personData != null) {
            data.put("person", personData);
        }
        data.put("client", buildClientData());
        data.put("notifier", buildNotifierData());

        return data;
    }

    private JSONObject buildPayload(JSONArray data) throws JSONException {
        JSONObject payload = new JSONObject();

        payload.put("access_token", this.accessToken);
        payload.put("data", data);

        return payload;
    }

    private JSONArray loadItems(File file) {
        Log.d(Rollbar.TAG, "Loading items...");

        JSONArray items = null;

        try {
            FileInputStream in = new FileInputStream(file);

            StringBuilder content = new StringBuilder();

            byte[] buffer = new byte[1024];
            while (in.read(buffer) != -1) {
                content.append(new String(buffer));
            }

            in.close();

            items = new JSONArray(content.toString());

            Log.d(Rollbar.TAG, "Items loaded.");
        } catch (FileNotFoundException e) {
            Log.e(Rollbar.TAG, "Unable to read item file.", e);
        } catch (IOException e) {
            Log.e(Rollbar.TAG, "Unable to read item file.", e);
        } catch (JSONException e) {
            Log.e(Rollbar.TAG, "Invalid item data. Deleting file.", e);
            file.delete();
        }

        return items;
    }

    public File writeItems(JSONArray items) {
        Log.d(Rollbar.TAG, "Writing items...");

        try {
            String filename = itemCounter++ + "." + System.currentTimeMillis();
            File file = new File(queuedItemDirectory, filename);
            FileWriter writer = new FileWriter(file);

            writer.write(items.toString());
            writer.close();

            Log.d(Rollbar.TAG, "Items written");
            return file;
        } catch (IOException e) {
            Log.e(Rollbar.TAG, "Unable to write items.", e);
            return null;
        }
    }

    public void postItems(final JSONArray items, final File file) {
        Log.i(Rollbar.TAG, "Sending item batch...");

        JSONObject payload;
        try {
            payload = buildPayload(items);
        } catch (JSONException e) {
            Log.e(Rollbar.TAG, PAYLOAD_ERROR_MSG, e);
            return;
        }

        HttpRequestManager.getInstance().postJson(this.endpoint, payload, false,
                new HttpResponseHandler() {

            @Override
            public void onSuccess(HttpResponse response) {
                Log.i(Rollbar.TAG, "Success");

                if (file != null) {
                    file.delete();
                }
            }

            @Override
            public void onFailure(HttpResponse response) {
                Log.e(Rollbar.TAG, "There was a problem reporting to Rollbar.");
                Log.e(Rollbar.TAG, "Response: " + response);

                // if the failed response has a 4xx status code, delete the file if it exsits
                if (response.hasStatusCode() && response.has4xxStatusCode()) {
                    if (file != null) {
                        file.delete();
                    }
                // otherwise, write a file if one doesn't exist already and reschedule
                } else {
                    if (file == null) {
                        writeItems(items);
                    }

                    scheduleItemFileHandler(itemScheduleDelay);
                }
            }
        });
    }

    private void scheduleItemFileHandler(int delay) {
        if (!handlerScheduled) {
            handlerScheduled = true;

            Log.d(Rollbar.TAG, "Scheduling item file handler...");

            scheduler.schedule(new Runnable() {

                @Override
                public void run() {
                    Log.d(Rollbar.TAG, "Item file handler running...");

                    File[] files = queuedItemDirectory.listFiles();

                    for (File file : files) {
                        JSONArray items = loadItems(file);
                        postItems(items, file);
                    }

                    handlerScheduled = false;
                    Log.d(Rollbar.TAG, "Item file handler finished.");
                }
            }, delay, TimeUnit.SECONDS);
        }
    }

    private JSONObject createTrace(Throwable throwable, String description) throws JSONException {
        JSONObject trace = new JSONObject();

        JSONArray frames = new JSONArray();

        StackTraceElement[] elements = throwable.getStackTrace();
        for (int i = elements.length - 1; i >= 0; --i) {
            StackTraceElement element = elements[i];

            JSONObject frame = new JSONObject();

            frame.put("class_name", element.getClassName());
            frame.put("filename", element.getFileName());
            frame.put("method", element.getMethodName());

            if (element.getLineNumber() > 0) {
                frame.put("lineno", element.getLineNumber());
            }

            frames.put(frame);
        }

        try {
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            PrintStream ps = new PrintStream(baos);

            throwable.printStackTrace(ps);
            ps.close();
            baos.close();

            trace.put("raw", baos.toString("UTF-8"));
        } catch (Exception e) {
            Log.e(Rollbar.TAG, "Exception printing stack trace.", e);
        }

        JSONObject exceptionData = new JSONObject();
        exceptionData.put("class", throwable.getClass().getName());
        exceptionData.put(MESSAGE, throwable.getMessage());

        if (!TextUtils.isEmpty(description)) {
        	exceptionData.put("description", description);
        }

        trace.put("frames", frames);
        trace.put("exception", exceptionData);

        return trace;
    }

    public void uncaughtException(Throwable throwable) {
        if (reportUncaughtExceptions) {
            Log.d(Rollbar.TAG, "Handling uncaught exception...");

            // Finish rollbar thread
            rollbarThread.interrupt();

            JSONObject item = buildItemPayload(throwable, uncaughtExceptionLevel, null);

            if (item != null) {
                JSONArray items = new JSONArray();
                items.put(item);

                if (sendOnUncaughtException) {
                    postItems(items, null);
                } else {
                    writeItems(items);
                }
            }

            try {
                // Try finishing up any items currently being sent
                rollbarThread.join();
            } catch (InterruptedException e) {
                Log.d(Rollbar.TAG, "Couldn't join rollbar thread", e);
            }
        }
    }

    private JSONObject buildItemPayload(Throwable throwable, String level, String description) {
        return buildItemPayload(throwable, level, description, new HashMap<String, String>());
    }

    private JSONObject buildItemPayload(Throwable throwable, String level, String description, Map<String, String> params) {
        try {
            JSONObject body = new JSONObject();

            List<JSONObject> traces = new ArrayList<JSONObject>();
            do {
                traces.add(0, createTrace(throwable, description));
                throwable = throwable.getCause();
            } while (throwable != null);
            JSONArray jsonArray = new JSONArray(traces);

            body.put("trace_chain", jsonArray);

            if (level == null) {
                level = defaultCaughtExceptionLevel;
            }

            for(Map.Entry<String, String> stringEntry : params.entrySet()){
                body.put(stringEntry.getKey(), stringEntry.getValue());
            }

            return buildData(level, body);
        } catch (JSONException e) {
            Log.e(Rollbar.TAG, PAYLOAD_ERROR_MSG, e);
            return null;
        }
    }

    private JSONObject buildItemPayload(String message, String level) {
        return buildItemPayload(message, level, new HashMap<String, String>());
    }

    private JSONObject buildItemPayload(String message, String level, Map<String, String> params) {
        try {
            JSONObject body = new JSONObject();
            JSONObject messageBody = new JSONObject();

            messageBody.put("body", message);
            body.put(MESSAGE, messageBody);

            for(Map.Entry<String, String> stringEntry : params.entrySet()){
                messageBody.put(stringEntry.getKey(), stringEntry.getValue());
            }
            return buildData(level, body);
        } catch (JSONException e) {
            Log.e(Rollbar.TAG, PAYLOAD_ERROR_MSG, e);
            return null;
        }
    }

    public void reportException(Throwable throwable, String level, String description) {
        JSONObject item = buildItemPayload(throwable, level, description);

        if (item != null) {
            rollbarThread.queueItem(item);
        }
    }

    public void reportException(Throwable throwable, String level, String description, Map<String, String> params) {
        JSONObject item = buildItemPayload(throwable, level, description, params);

        if (item != null) {
            rollbarThread.queueItem(item);
        }
    }

    public void reportMessage(String message, String level) {
        JSONObject item = buildItemPayload(message, level);

        if (item != null) {
            rollbarThread.queueItem(item);
        }
    }

    public void reportMessage(String message, String level, Map<String, String> params) {
        JSONObject item = buildItemPayload(message, level, params);

        if (item != null) {
            rollbarThread.queueItem(item);
        }
    }

    public void setPersonData(JSONObject personData) {
        this.personData = personData;
    }

    public void setPersonData(String id, String username, String email) {
        JSONObject personData = new JSONObject();

        try {
            personData.put("id", id);

            if (username != null) {
                personData.put("username", username);
            }
            if (email != null) {
                personData.put("email", email);
            }

            this.personData = personData;
        } catch (JSONException e) {
            Log.e(Rollbar.TAG, "JSON error creating person data.", e);
        }
    }

    public void setEndpoint(String endpoint) {
        this.endpoint = endpoint;
    }

    public void setReportUncaughtExceptions(boolean reportUncaughtExceptions) {
        this.reportUncaughtExceptions = reportUncaughtExceptions;
    }

    public void setIncludeLogcat(boolean includeLogcat) {
        this.includeLogcat = includeLogcat;
    }

    public void setDefaultCaughtExceptionLevel(String level) {
        this.defaultCaughtExceptionLevel = level;
    }

    public void setUncaughtExceptionLevel(String level) {
        this.uncaughtExceptionLevel = level;
    }

    public void setSendOnUncaughtException(boolean sendOnUncaughtException) {
        this.sendOnUncaughtException = sendOnUncaughtException;
    }

    public void setItemScheduleDelay(int delay) {
        this.itemScheduleDelay = delay;
    }
}
